pkg/api/api.go

```diff
@@ -123,28 +68,54 @@ var AllowedClientSigningAlgorithms = []v1.PublicKeyDetails{
 }
 var DefaultClientSigningAlgorithms = AllowedClientSigningAlgorithms
 
-func NewAPI(treeID uint) (*API, error) {
-	logRPCServer := fmt.Sprintf("%s:%d",
-		viper.GetString("trillian_log_server.address"),
-		viper.GetUint("trillian_log_server.port"))
+func NewAPI(treeID int64) (*API, error) {
 	ctx := context.Background()
-	tConn, err := dial(logRPCServer)
-	if err != nil {
-		return nil, fmt.Errorf("dial: %w", err)
+
+	// this is also used for the active tree
+	defaultGRPCConfig := trillianclient.GRPCConfig{
+		Address:             viper.GetString("trillian_log_server.address"),
+		Port:                viper.GetUint16("trillian_log_server.port"),
+		TLSCACert:           viper.GetString("trillian_log_server.tls_ca_cert"),
+		UseSystemTrustStore: viper.GetBool("trillian_log_server.tls"),
 	}
-	logAdminClient := trillian.NewTrillianAdminClient(tConn)
-	logClient := trillian.NewTrillianLogClient(tConn)
 
-	tid := int64(treeID)
-	if tid == 0 {
+	if treeID == 0 {
 		log.Logger.Info("No tree ID specified, attempting to create a new tree")
-		t, err := trillianclient.CreateAndInitTree(ctx, logAdminClient, logClient)
+		t, err := trillianclient.CreateAndInitTree(ctx, defaultGRPCConfig, viper.GetString("trillian_log_server.grpc_default_service_config"))
 		if err != nil {
 			return nil, fmt.Errorf("create and init tree: %w", err)
 		}
-		tid = t.TreeId
+		treeID = t.TreeId
+	}
+
+	shardingConfig := viper.GetString("trillian_log_server.sharding_config")
+	signingConfig := signer.SigningConfig{
+		SigningSchemeOrKeyPath: viper.GetString("rekor_server.signer"),
+		FileSignerPassword:     viper.GetString("rekor_server.signer-passwd"),
+		TinkKEKURI:             viper.GetString("rekor_server.tink_kek_uri"),
+		TinkKeysetPath:         viper.GetString("rekor_server.tink_keyset_path"),
+		GCPKMSRetries:          viper.GetUint("rekor_server.signer.gcpkms.retries"),
+		GCPKMSTimeout:          viper.GetUint("rekor_server.signer.gcpkms.timeout"),
+	}
+	ranges, err := sharding.NewLogRanges(ctx, shardingConfig, treeID, signingConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unable get sharding details from sharding config: %w", err)
+	}
+
+	grpcConfigs := make(map[int64]trillianclient.GRPCConfig)
+	for _, r := range ranges.GetInactive() {
+		if r.GRPCConfig != nil {
+			grpcConfigs[r.TreeID] = *r.GRPCConfig
+		}
+	}
+	tcm := trillianclient.NewClientManager(grpcConfigs, defaultGRPCConfig, viper.GetString("trillian_log_server.grpc_default_service_config"))
```

Comment:
Should `grpc_default_service_config` be a part of a trillianclient.config struct?

Reply:
I could make this similar to address/port/etc, thanks for the suggestion